extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/LDAPConnectionService.java

```diff
@@ -111,6 +113,29 @@ public LDAPConnection bindAs(String userDN, String password)
         // Obtain appropriately-configured LDAPConnection instance
         LDAPConnection ldapConnection = createLDAPConnection();
 
+        // Configure LDAP connection constraints
+        LDAPConstraints ldapConstraints = ldapConnection.getConstraints();
+        if (ldapConstraints == null)
+          ldapConstraints = new LDAPConstraints();
+
+        // Set whether or not we follow referrals
+        ldapConstraints.setReferralFollowing(confService.getFollowReferrals());
+
+        // If the referral auth method is set to bind, we set it using the existing
+        // username and password.
+        String refAuthMethod = confService.getReferralAuthentication();
+        if (refAuthMethod != null && refAuthMethod.equals("bind"))
+            ldapConstraints.setReferralHandler(new ReferralAuthHandler(userDN, password));
```

Comment:
So, what's the idea behind configuring whether referrals auth with the same credentials?

Reply:
I believe it's the difference between using the credentials that the user logged in with vs. using the credentials specified in the guacamole.properties file, if any.  So, in some cases it'll be the difference between an anonymous bind and an authenticated bind, in other cases it'll be the difference between binding with the configured search user vs. the logged in user.

There are cases where the user who logged in may not have access to the entire tree, or even the base of the tree, so you'd want the search & referral following to happen under that user account instead of the logged in account.

At least, that's the theory, I think...whether or not I've actually implemented it that way, I probably should go back and take another look and make sure.

Reply:
Okay, I configured it to just use the search credentials.

The one situation that I thought about here that you might want the ability to configure separate referral credentials is where you know you have a situation where your base LDAP tree points to another system, and the credentials are different on the referred to system than they are on the base system.  This is somewhat of a corner case, and I think it unlikely there are just huge numbers of people running that, but it's worth keeping in mind for future discussions.  In that case, however, just having this parameter - ldap-referral-auth - wouldn't be enough - you'd also want to ability to configure a separate username and password for the referrals, so you'd end up adding a couple more parameters.

Reply:
Well, to clarify a bit, there are two types of searches which may occur in the LDAP auth:

* The search performed to resolve the user's DN. This is the only case where the search DN and password are used.
* The search to retrieve the users and/or connections defined within LDAP and accessible to the user that authenticated. These searches naturally **must** use the user's own DN and password, since part of the point of the LDAP auth is to allow the LDAP directory's own access control to take effect.

If users or connections cannot be retrieved via LDAP because the authenticated user cannot access them, then they shouldn't have access, referral or not.

Reply:
Yeah, I see what you're saying.  The changes here should be consistent with that - the referral authentication just uses what is passed in to the bindAs method for userDN and password, which, for the initial user search will be the search credentials, and subsequently will be the retrieved userDN and the password used for authentication.

Reply:
Ah, OK then.